src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java

```diff
@@ -170,31 +171,69 @@ public void close() {
     ) {
         // if it's a starting clause
         if (this._child == null) {
-            // create initial RDD from expression
-            JavaRDD<Item> initialRdd = _expression.getRDD(context);
+            return getDataFrameFromRDDExpression(context);
+        }
 
-            // define a schema
-            List<StructField> fields = new ArrayList<>();
-            StructField field = DataTypes.createStructField(_variableName, DataTypes.BinaryType, true);
-            fields.add(field);
-            StructType schema = DataTypes.createStructType(fields);
+        if (_child.isDataFrame()) {
+            if (_expression.isRDD()) {
+                Set<String> intersection = new HashSet<>(_expression.getVariableDependencies().keySet());
+                intersection.retainAll(getVariablesBoundInCurrentFLWORExpression());
+                boolean expressionUsesVariablesOfCurrentFlwor = !intersection.isEmpty();
+
+                if (expressionUsesVariablesOfCurrentFlwor) {
+                    throw new InvalidForClauseExpressionException(
+                            "An RDD-enabled for clause expression cannot reference the variables defined in the encapsulating, RDD-enabled FLWOR expression.",
```

Comment:
I suggest this message, because RDD-enabled expressions is implementation terminology that we want to shield users from (some of our users are not computer scientists and do not know the details of Spark). The terminology we typically use for the end user is the notion of "big sequence" of items.

"A for clause expression cannot produce a big sequence of items for a big number of tuples, as this would lead to a data flow explosion."

What do you think?

Reply:
Sounds great. Thanks for the suggestion